lib/elixir/test/elixir/file_test.exs

```diff
@@ -21,6 +21,328 @@ defmodule FileTest do
+      end
+    end
+
+    test :rename_parent_dir_to_non_existing_sub_dir do
+      src  = tmp_fixture_path("cp_r")
+      dest = tmp_path("cp_r/x")
+      try do
+        assert File.exists?(src)
+        assert File.rename(src, dest) == {:error, :einval}
+        assert File.exists?(src)
+        refute File.exists?(dest)
+      after
+        File.rm_rf src
+        File.rm_rf dest
+      end
+    end
+
+    test :rename_dir_to_existing_empty_dir_default_behaviour do
+      src  = tmp_fixture_path("cp_r")
+      dest = tmp_path("tmp")
+
+      File.mkdir(dest)
+
+      try do
+        refute File.exists?(tmp_path("tmp/a"))
+
+        assert File.rename(src, dest) == :ok
+        {:ok, files} = File.ls(dest)
+        assert length(files) == 2
+        assert "a" in files
+
+        {:ok, files} = File.ls(tmp_path("tmp/a"))
+        assert length(files) == 2
+        assert "1.txt" in files
+
+        assert File.exists?(tmp_path("tmp/a/1.txt"))
+        assert File.exists?(tmp_path("tmp/a/a/2.txt"))
+        assert File.exists?(tmp_path("tmp/b/3.txt"))
+
+        refute File.exists?(src)
+      after
+        File.rm_rf src
+        File.rm_rf dest
+      end
+    end
+
+    test :rename_dir_to_existing_empty_dir do
+      src  = tmp_fixture_path("cp_r")
+      dest = tmp_path("tmp")
+
+      File.mkdir(dest)
+
+      try do
+        assert File.exists?(dest)
+        assert File.rename(src, dest) == :ok
+        refute File.exists?(src)
+        assert File.exists?(tmp_path("tmp/a"))
+      after
+        File.rm_rf src
+        File.rm_rf dest
+      end
+    end
+
+    test :rename_dir_to_existing_non_empty_dir do
+      src  = tmp_fixture_path("cp_r")
+      dest = tmp_path("tmp")
+
+      File.mkdir_p(tmp_path("tmp/x"))
+
+      try do
+        assert File.exists?(tmp_path("tmp/x"))
+        assert File.exists?(src)
+        refute File.exists?(tmp_path("tmp/a"))
+
+        assert File.rename(src, dest) == {:error,:eexist}
+
+        assert File.exists?(tmp_path("tmp/x"))
+        assert File.exists?(src)
+        refute File.exists?(tmp_path("tmp/a"))
+      after
+        File.rm_rf src
+        File.rm_rf dest
+      end
+    end
+
+    test :rename_unknown_source do
+      src  = fixture_path("unknown")
+      dest = tmp_path("tmp")
+      try do
+        assert File.rename(src, dest) == {:error, :enoent}
+      after
+        File.rm_rf dest
+      end
+    end
+
+    test :rename_preserves_mode do
+     File.mkdir_p!(tmp_path("tmp"))
+     src  = tmp_fixture_path("cp_mode")
+     dest = tmp_path("tmp/cp_mode")
+
+     try do
+       %File.Stat{mode: src_mode} = File.stat! src
+       File.rename(src, dest)
+       %File.Stat{mode: dest_mode} = File.stat! dest
+       assert src_mode == dest_mode
+      after
+        File.rm_rf src
+        File.rm_rf dest
+      end
+    end
+
+    def tmp_fixture_path(extra) do
+      src = fixture_path(extra)
+      dest = tmp_path(extra)
+      File.cp_r(src,dest)
+      dest
+    end
+
```

Comment:
That's my editor; I like the newline incase I need to cat the file but I don't like it enough to argue it's merits :-)